### DISCUSSION TITLE
No network route in the hello.c example topology? how does it find the route?

### DISCUSSION THREAD

**Discussion by GitZinger** [shadow/shadow#767] [Q&A] [answered]
**State your question concisely**
https://github.com/shadow/shadow-plugin-extras/blob/master/hello/example.xml
In the hello.c example topology, the server and client nodes are not indicated in the **topology**(in it there is only one node called **poi-1**, how does this poi-1 **associate with** those two node id **helloserver and helloclient** ), besides, Is it consistent with the newest version of Shadow? because in the newest Mini example, it uses the **host id="xxx"** not **node id="xxx"**
neither does the edge between the server and client nodes? then how do they communicate with each other?  
if no edge is needed. then the topology is useless?

**Describe your current understanding of Shadow as it relates to your question**
I thought we need to specify all the route between the nodes to let them communicate with each other?
**Describe which parts you are unsure about**
have no clue how the network layer communication works?
please advise me!

**Comment by robgjansen** [comment, marked as answer]
In the topology xml, vertices represent places where you can attach hosts into the network. The topology vertices (i.e., the node id poi-1 in your example) are described here:
https://github.com/shadow/shadow/blob/master/docs/3.2-Network-Config.md#vertices

> Vertices
>
> In addition to the default id attribute, all vertices must have the bandwidthup and bandwidthdown attributes (given in KiB/s). Optional vertex attributes include ip, citycode, countrycode, and packetloss.
>
> You can control where hosts in your Shadow experiment get attached to the topology by setting the iphint, citycodehint, and countrycodehint in your host attribute list in the Shadow config file. Hosts are always attached to the closest match to the best vertex following the hinted restrictions.

So you can add `ip`, `citycode`, and/or `countrycode` attributes to your poi-1 node to specify where that node is located.

The next step is to figure out how to assign hosts onto those topology vertices. If you want to control where a specific Shadow hosts gets assigned in the network topology, you should use the `iphint`, `citycodehint`, and/or `countrycodehint` attributes to the host element in shadow.config.xml.

The iphint, citycodehint, and countrycodehint attributes provide hints to Shadow's name and routing system on where to attach this host to the topology. These hints will be matched based on the values of the ip, citycode, and countrycode of the vertices as specified in the topology file.

If the hints are not provided on the hosts (like they are not in your example), then Shadow will randomly assign the host to a topology vertex. In your example, there is only one single topology vertex, so both hosts will get assigned to it, and the latency between the hosts are given by the single edge in the topology.
